### PR TITLE
Move DifferentiationInvoker and ADContext to separate files.

### PR DESCRIPTION
**Pull request by bgogul**
This PR is part of refactoring Differentiation.cpp (https://bugs.swift.org/browse/TF-993)

### CODE REVIEW

**Comment by bgogul** [comment]
@swift-ci please test tensorflow

File: lib/SILOptimizer/Utils/Differentiation/ADContext.cpp [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+
+using llvm::DenseMap;
+using llvm::SmallPtrSet;
+using llvm::SmallVector;
+
+namespace swift {
+namespace autodiff {
+
+//===----------------------------------------------------------------------===//
+// Local Helpers
```
Comment (dan-zheng) [inline lib/SILOptimizer/Utils/Differentiation/ADContext.cpp]: (minor, feel free to address later)

Please capitalize only the first word in headings, for consistency with headings in other files.
```suggestion
// Local helpers
```

File: lib/SILOptimizer/Utils/Differentiation/ADContext.cpp [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+    if (!fd || !fd->isStatic() || !fd->isOperator())
+      continue;
+    return fd;
+  }
+  // Not found.
+  return nullptr;
+}
+
+//===----------------------------------------------------------------------===//
+// ADContext methods.
```
Comment (dan-zheng) [inline lib/SILOptimizer/Utils/Differentiation/ADContext.cpp]: (minor, feel free to address later)

No period after headings, for consistency with headings in other files.
```suggestion
// ADContext methods
```

File: include/swift/SILOptimizer/Utils/Differentiation/ADContext.h [hunk 3]
```diff
@@ -0,0 +1,349 @@
...
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+//
+// SWIFT_ENABLE_TENSORFLOW
+//
+// Per-module contextual information for the Differentiation pass.
+//===----------------------------------------------------------------------===//
```
Comment (dan-zheng) [inline include/swift/SILOptimizer/Utils/Differentiation/ADContext.h]: (minor, feel free to address later)

Empty line between file header comment and code please.
```suggestion
//===----------------------------------------------------------------------===//

```

**Comment by bgogul** [comment]
@swift-ci please test tensorflow

**Review by dan-zheng — APPROVED** [review]
Yay for separation!

**Review by rxwei — APPROVED** [review]
Thank you!

**Pull request merged**